lib/devise/hooks/two_step_verification.rb

```diff
@@ -1,5 +1,8 @@
 Warden::Manager.after_authentication do |user, auth, _options|
   if user.respond_to?(:need_two_step_verification?)
-    auth.session(:user)['need_two_step_verification'] = user.need_two_step_verification?
+    cookie = auth.env["action_dispatch.cookies"].signed["remember_2sv_session"]
+    unless cookie && cookie[:user_id] == user.id && cookie[:expires] > Time.now
```

Comment:
I'd be tempted to pull this line and the line above out into a method called something like `has_valid_2sv_cookie?`